Document undocumented options

SkySkimmer opened the pull request:
List of yet to be done options: (`rm 8.8` annotations are not exhaustive)
```
Debug Ltac
Debug Ssreflect
Debug SsrMatching
Debug Tactic Unification
Debug Typeclasses
Debug Unification
Default Clearing Used Hypotheses
Default Proof Mode
Dependent Propositions Elimination (rm 8.8)
Discriminate Introduction (rm 8.8)
Dump Bytecode
Extraction AccessOpaque
Extraction File Comment
Extraction Flag
Extraction name
Extraction TypeExpand
Functional Induction Rewrite Dependent
Function_debug
Function_raw_tcc
Injection L2R Pattern Order
Inline Level
Kernel Term Sharing
Keyed Unification
Lia Depth
Lia Enum
Lra Depth
NativeCompute Profile Filename
Omega OldStyle
Printing Goal Names
Printing Goal Tags
Program Mode
Rtauto Check
Rtauto Pruning
Rtauto Verbose
SimplIsCbn
Solve Unification Constraints
SsrHave NoTCResolution
SsrIdents
SsrMatchingProfiling
SsrOldRewriteGoalsOrder
SsrProfiling
SsrRewrite
Standard Proposition Elimination Names (rm 8.8)
Tactic Pattern Unification (rm 8.8)
Typeclasses Depth
Typeclasses Iterative Deepening
```
The list was obtained by grepping for `optkey =` and postprocessing a bit (notably a few places use a function to produce the option record, so you get `optkey = some variable`), grepping for `\optindex{}` and postprocessing a bit, then diffing the results.

# Experimental or otherwise intentionally undocumented options

"intentionally undocumented" is pretty suspicious TBH.
For instance:
- `Proof Using Clear Unused` which has commented-out doc in the .tex
- `Default Clearing Used Hypotheses` introduced by @herbelin 262e3151ce483aaab3b4f60e3d1dbdf875ea05ae (2014)
That commit also introduced a syntax like `destruct !x` which doesn't seem to exist any more. Is this just a forgotten option? It still works though.

# Additional notes:
- `Set Firstorder Solver` is in the option index but does not use the goptions mechanism since it takes a `tactic` value. It uses `Tactic_option` which is also used by `Obligation Tactic :=`, which is not in the optindex. Should it be? Should (and can) goptions be extended to subsume tactic_option?
EDIT: Declare Implicit Tactic has a similar purpose but rolls its own libobject calls.
- several of these options are deprecated with a comment `remove in 8.8` or the like. Is it worth documenting them? I guess yes if only to backport in 8.7.
- several of these options are debug options. Is it worth documenting them?

SkySkimmer commented:
Everyone please feel free to directly push documentation commits to my branch.
If you want to avoid work duplication simply edit the head post to add a `*` to options you're working on, I will also do that.

ppedrot commented:
Note that #6164 is included in this issue.

Zimmi48 commented:
Thanks @SkySkimmer, great initiative!

> several of these options are deprecated with a comment remove in 8.8 or the like. Is it worth documenting them? I guess yes if only to backport in 8.7.

I would prefer that documentation of these deprecated options be rather done in a separate 8.7-only PR so as to not interfere with PRs removing such options, like #6169.

> several of these options are debug options. Is it worth documenting them?

Yes, I guess. At least in the developer's documentation.

SkySkimmer commented on doc/refman/RefMan-pro.tex:
```diff
@@ -103,6 +103,14 @@ \section{Switching on/off the proof editing mode}
+variables in the context (or by those in the {\tt Proof Using}
+expression if it exists) at the end of the section. Unsetting this
+option generalizes {\tt Admitted} lemmas by only those variables used
+in its type and partial proof.
```
This seems like a corner case where async might do the wrong thing.
OTOH I think it's disabled for Admitted.

maximedenes replied:
This documentation doesn't seem to be reflecting the current behavior:
```
Unset Keep Admitted Variables.

Section Foo.

Variable x : nat.

Lemma foo : True.
Proof.
generalize x.
Show Proof. (* x is used in the partial term *)
Admitted.
End Foo.
About foo. (* no hypothesis *)
```

Please fix or clarify if it is a bug.

SkySkimmer replied:
I forgot what tests I did to figure out what this option does so I'm just dropping it.

SkySkimmer commented:
`Omega Oldstyle` controls a `bool ref` which has been around since 44d578b40ec699ea8bbd4b387c2cf7155bf1d1fe (omega into CVS, 2000).
I have no idea what it does.

SkySkimmer commented:
I can't find any documentation for unification and unification constraints, so I don't know where `Solve Unification Constraints` would go.

JasonGross commented on doc/refman/RefMan-mod.tex:
```diff
@@ -403,10 +403,14 @@ \subsection{\tt Import {\qualid}
 Prints the module type and (optionally) the body of the module {\ident}.
 
+For this command and {\tt Print Module Type}, the option {\tt Short
+  Module Printing} disables the printing of the types of fields,
```
Is it on or off by default?

SkySkimmer replied:
Off, and may as well.

JasonGross commented on doc/refman/RefMan-tac.tex:
```diff
@@ -3334,6 +3334,14 @@ \subsection{{\tt cbv \flag$_1$ \mbox{\dots} \flag$_n$}, {\tt lazy \flag$_1$
 
 \end{Variants}
 
+\Rem The following option makes {\tt cbv} (and its derrivative {\tt
```
`derrivative` -> `derivative`

JasonGross commented on pretyping/coercion.ml:
```diff
@@ -36,7 +36,7 @@ let _ =
     { optdepr  = false;
       optname  = "use typeclass resolution during conversion";
-      optkey   = ["Typeclass"; "Resolution"; "For"; "Conversion"];
+      optkey   = ["Typeclasses"; "Resolution"; "For"; "Conversion"];
```
Really?  I thought it's always "typeclass resolution" (to refer to the resolution used by all typeclasses), in the same way that one might say "gamer culture" to refer to the culture of all gamers.

SkySkimmer replied:
I made it match the already existing documentation and the other typeclass options.
It's a component prefix, like `Typeclasses.resolution_for_conversion`.

JasonGross commented on tactics/tactics.ml:
```diff
@@ -87,7 +87,7 @@ let _ =
     { optdepr = true; (* remove in 8.8 *)
       optname = "Perform typeclass resolution on apply-generated subgoals.";
-      optkey = ["Typeclass";"Resolution";"After";"Apply"];
+      optkey = ["Typeclasses";"Resolution";"After";"Apply"];
```
As above

JasonGross commented:
I may write up a couple of the ones that I understand.  Also, cc @tchajed

herbelin commented:
> "intentionally undocumented" is pretty suspicious TBH.
> For instance:
> - `Proof Using Clear Unused` which has commented-out doc in the .tex

See #883 for a pending discussion related to this option.

> - `Default Clearing Used Hypotheses` introduced by @herbelin 262e3151ce483aaab3b4f60e3d1dbdf875ea05ae (2014)
> That commit also introduced a syntax like `destruct !x` which doesn't seem to exist any more. Is this just a forgotten option? It still works though.

The syntax `destruct !x` has been replaced by `destruct (x)` on a suggestion from @silene as part of an attempt to harmonize the syntax of ssreflect and vanilla coq.

The syntaxes `apply >H`, `rewrite >H`, `elim >H`, etc. exist but are undocumented (they remove the hypothesis on the fly). I think that there are useful and should be documented. My idea was that the option `Default Clearing Used Hypotheses` becomes at some time the default behavior, so that, by default, `apply` and `rewrite` clear the hypothesis they use, based on the experience that it is more common to use an hypothesis once than strictly more than once.

I let this pending because of the tensions with ssreflect developers (see e.g the` %pat` story, and the various not-always-very-fair criticisms at this time). At some time, I thought the only way to get out of the tensions was to bring ssreflect developers at a table to work at a consensual project about tactics. I would like to be shown wrong but I tend to lose hope that this is going to happen. So I would vote so that vanilla coq continues to evolve on its own, and, when the versioning of tactics is operational, that the option `Default Clearing Used Hypotheses` is documented and (a priori) better set on by default. It is certainly not a fundamental change but still a convenient one.

maximedenes commented:
I'm cancelling the Travis build to reduce the backlog, as I think this needs more work (and discussion maybe). I hope it's ok.

Zimmi48 commented:
Similarly to @ppedrot remark, note that documenting `Set Keyed Unification` would fix #4551.

Zimmi48 commented:
@SkySkimmer WDYT of merging this as-is to allow shipping this additional documentation in 8.7.1 and creating a new issue to track the reminding missing option documentation?
As for the two no-op option removals and the renaming of a third option to match the documentation, I have an uneasy feeling about including this in a patch-level release. Would you mind adding some deprecation warnings instead of removing them completely and submitting another PR for their definite removal? (Especially in the case of the renaming.)

SkySkimmer commented:
Done.

maximedenes commented:
I would like that we discuss a bit if we don't want to remove some of these options instead of documenting them.

Zimmi48 commented:
@maximedenes I'm unclear about your approach. Removing an option generally has to go through a phase of deprecation. That's orthogonal to the question of documentation. In fact, I'm very suspicious about the general tradition in the Coq Development Team of not documenting because we are not sure it will stay. When that's the case, just document as "experimental, could be removed without a warning". Similarly, I don't see what good can come out of blocking a documentation PR.

maximedenes commented:
I agree that introducing undocumented options is suspicious and should be avoided. . But on the other hand, documenting existing options that were not is clearly a signal that they can be used. So I'd like to have a bit of time, at least to calmly go through the list. I'm very busy these days, sorry for the delay.

tchajed commented on doc/refman/Cases.tex:
```diff
@@ -280,6 +280,19 @@
 \end{coq_example}
 
+The option {\tt Set Asymmetric Patterns} \optindex{Asymmetric Patterns}
+(off by default) removes the parameters from the pattern
```
"removes parameters to constructors in patterns" would be more clear to me, with less need for all the preceding context

tchajed commented on doc/refman/Omega.tex:
```diff
@@ -149,6 +149,32 @@ \section{Using {\tt omega}}
+  {\tt Unset Stable Omega}
+\end{quote}
+This deprecated option (on by default) is for compatibility with Coq
+pre 8.5. It resets internal name counters to make iterated executions
```
I think you can just remove "iterated"

tchajed commented on plugins/firstorder/g_ground.ml4:
```diff
@@ -40,17 +40,17 @@ let _=
+  let congruence_depth=ref 100
   let gdopt=
-    { optdepr=false;
+    { optdepr=true; (* noop *)
```
I'm not sure why this is in a documentation PR - shouldn't deprecating options be separate from documenting them?

tchajed commented on plugins/ssr/ssrequality.ml:
```diff
@@ -141,16 +141,16 @@ let newssrcongrtac arg ist gl =
       Goptions.optkey   = ["Match"; "Strict"];
       Goptions.optread  = (fun () -> !ssr_strict_match);
-      Goptions.optdepr  = false;
+      Goptions.optdepr  = true; (* noop *)
```
Same here - deprecation should be separate from documentation as far as I'm concerned.

tchajed commented on tactics/tactics.ml:
```diff
@@ -83,6 +83,14 @@ let _ =
+  declare_bool_option
+    { optdepr = true; (* remove in 8.8 *)
+      optname = "Perform typeclass resolution on apply-generated subgoals.";
+      optkey = ["Typeclasses";"Resolution";"After";"Apply"];
```
First of all, not sure why we're adding an option here.
Second, it should be called "Typeclass Resolution After Apply".

Zimmi48 replied:
In fact, the option is added for compatibility (it already existed as such but was documented as "Typeclass Resolution After Apply".

tchajed commented:
@maximedenes I think if an option is accepted by Coq, it should be documented; it's ok to merge (accurate) documentation even if the option is soon removed, though the documentation should probably call out the option as being unstable/subject to removal.

If I understood correctly you're saying documenting an option blesses it as being acceptable to use. This is a scary idea to me: if I'm able to use an option, I consider it part of Coq and am frustrated when compatibility breaks. When I see an undocumented option, I consider it a documentation bug. If there are private, internal options, at the very least they should be documented as being internal, and ideally their use would be discouraged by Coq itself (by issuing a warning or requiring a flag to enable them).

JasonGross commented on doc/refman/RefMan-pro.tex:
```diff
@@ -103,6 +103,14 @@ \section{Switching on/off the proof editing mode}
 
+\variant {\tt Unset Keep Admitted Variables} \optindex{Keep Admitted Variables}
+
+By default in sections, an {\tt Admitted} lemma is generalized by all
```
Wording nit: "generalized over all" instead of "generalized by add"

Zimmi48 replied:
@SkySkimmer You didn't address this comment :) but then I think I bothered you enough on this PR already...

SkySkimmer replied:
That's a github illusion, I removed that commit.

SkySkimmer replied:
Or maybe it didn't get removed for some reason? It's gone now.

JasonGross commented on doc/refman/RefMan-tac.tex:
```diff
@@ -3507,6 +3515,12 @@ \subsection{\texorpdfstring{\texttt{cbn}}{cbn} and \texorpdfstring{\texttt{simpl
+  \optindex{Debug RAKAM}
+  {\tt Set Debug RAKAM}
+\end{quote}
+This option makes {\tt cbn} print various debugging information.
```
Non-blocking: Might be nice to say what `RAKAM` stands for (r... a... k... abstract machine?)

ppedrot replied:
I guess @pirbo has the answer. Maybe [a tribute to Tintin](https://en.wikipedia.org/wiki/Red_Rackham%27s_Treasure)?

maximedenes replied:
My guess is `R` probably stands for refolding. The first `A` I have no idea, though. So @ppedrot's explanation is more plausible.

maximedenes replied:
`K` is for Krivine.

pirbo replied:
Refolding Algebraic Krivine Abstract Machine

maximedenes replied:
@pirbo What does algebraic mean in this context?!

pirbo replied:
It means that it is a machine that does `iota` in addition of `beta``delta`.
Said differently, it deals with algebraic datatypes (via case analyses/(co)fixpoint reduction) which the Krivine Abstract Machine has no idea about.

It comes from the (only on archive (and in my thesis)) description of the system:
- take a standard KAM
- add the machinery to deal with algebraic data structure (AKAM)
- add a second stack to deal with unfolded constant (RAKAM)

You obtain a imitation of `simpl` at 95% that have some simplicity/efficiency/customazability advantages over `simpl`  ( <-- argument contested), can be used to make unification refold while it reduced or be used in a strait forward way by `hnf` to make it refold

JasonGross commented on doc/refman/RefMan-tac.tex:
```diff
@@ -4710,6 +4750,13 @@ \subsection{\tt congruence}
     described above.
 \end{ErrMsgs}
 
+\noindent {\bf Remark: } {\tt congruence} can be made to print debug
```
What's the difference between `\Rem` and `\noindent {\bf Remark: } `, and why choose one over the other?

SkySkimmer replied:
I have no idea, IIRC I just picked whichever I saw first (without caching past picks).

Zimmi48 replied:
Probably irrelevant anyways since we are soon moving to Sphinx.

JasonGross commented on pretyping/coercion.ml:
```diff
@@ -35,6 +35,15 @@ let use_typeclasses_for_conversion = ref true
+
+let _ =
+  Goptions.(declare_bool_option
+    { optdepr  = true; (* alias of see above, remove in 8.8 *)
```
I think the other one (`Typeclasses Resolution`) should be deprecated, and the documentation should be updated.

Zimmi48 replied:
The other one didn't exist before this PR IIRC. So what you want in fact is that the documentation be updated according to the existing options, instead of the option being renamed to match the documentation.

Zimmi48 commented:
@SkySkimmer WDYT should be done with this? Address the comments and force Maxime's hand? Or postpone? I'm personally favorable to the first option.

maximedenes commented on doc/refman/RefMan-tac.tex:
```diff
@@ -3703,6 +3717,9 @@ \subsection{\tt auto}
   This variant is very useful for getting a better understanding of automation,
   or to know what lemmas/assumptions were used.
 
+\item {\tt debug auto} Behaves like {\tt auto} but shows the tactics
```
how does it relate to ` info`, which IIRC is what people were using?

maximedenes replied:
Feel free to ignore the remark if this is written somewhere else, I don't have the time to read everything.

SkySkimmer replied:
>how does it relate to info, which IIRC is what people were using?

I don't know, I'm not much of a `auto` user.

Zimmi48 replied:
`debug` shows more, it also shows the dead branches, whereas `info` shows just the successful branch.

maximedenes commented:
I reviewed the commands that are being documented, it looks ok.

Please merge only the documentation part in 8.7.1, though (no deprecation, no renaming, no creation of new options).

Zimmi48 commented:
> Please merge only the documentation part in 8.7.1, though (no deprecation, no renaming, no creation of new options).

In that case, the PR needs to be updated before being merged. (And sorry @SkySkimmer for the work done twice.)

SkySkimmer commented:
I removed the renamings and fixed the optindex instead (it turns out that the documentation was accurate except for the index).
I moved the deprecations to make them the last commits, it should be trivial to avoid backporting them. Considering they're dead code options I would still deprecate them though.

Zimmi48 commented:
@maximedenes I agree with @SkySkimmer: I would usually refuse to backport deprecation notices in v8.7 (after the 8.7.0 release) but given that they are dead code options, I think it makes sense to remove them in master and thus to deprecate them in v8.7.
So this looks ready to be merged.

Zimmi48 commented:
@SkySkimmer I think your Travis indicates a compilation error. Can you check?

SkySkimmer commented:
Typo introduced syntax error
Fixed now

Zimmi48 commented:
Being built here: https://travis-ci.org/SkySkimmer/coq/builds/316557209
I suppose the error on Equations is just due to not being based on the latest version of master. Given the scope of the PR, it can be safely ignored, and the same should be true for the upcoming likely Ltac2 failure.

Zimmi48 commented:
There are lots of failures which all look spurious, we have at least a few successful test-suite runs and documentation builds.
